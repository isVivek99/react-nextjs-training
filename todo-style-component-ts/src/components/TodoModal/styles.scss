@import '../../sass/index.scss';

.todomodal {
  .form-error-text {
    font-size: 0.75rem;
    font-weight: bold;
    color: red;
  }
}
.modal-overlay {
  background-color: rgba(0, 0, 0, 0.5);
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  margin: auto;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}
.toastmodal_btn {
}
.toastmodal_btn_text {
  color: $color-white;
}
