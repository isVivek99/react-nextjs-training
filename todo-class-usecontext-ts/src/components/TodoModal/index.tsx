import React from 'react';
import { TodoContext, actions } from '../../store';
import './styles.scss';

interface TodoModalProps {
  onClose: () => unknown;
  actionBtnTitle: string;
  todo?: any;
  newTodo: boolean;
  index?: number;
}

interface Errors {
  titleError: string;
  descriptionError: string;
}
interface TodoModalState {
  title: string;
  description: string;
  errors: Errors;
}

export default class TodoModal extends React.Component<
  TodoModalProps,
  TodoModalState
> {
  state: TodoModalState = {
    title: '',
    description: '',
    errors: {
      titleError: '',
      descriptionError: '',
    },
  };

  handleSubmit = (e: any) => {
    e.preventDefault();

    // form validation
    if (
      this.state.title.trim() === '' &&
      this.state.description.trim() === ''
    ) {
      this.setState({
        errors: {
          titleError: 'Title is required',
          descriptionError: 'Description is required',
        },
      });
    } else if (this.state.title.trim() === '') {
      this.setState({ description: '', title: 'Title is required' });
    } else if (this.state.description.trim() === '') {
      this.setState({ description: 'Description is required', title: '' });
    } else {
      if (this.props.newTodo) {
        this.context.addTodo({
          title: this.state.title,
          description: this.state.description,
        });
        this.props.onClose();
      } else {
        this.props.onClose();
      }
    }
  };

  // useEffect(() => {
  //   setTitle(todo ? todo.title : '');
  //   setDescription(todo ? todo.description : '');
  // }, [todo]);
  static contextType = TodoContext;
  context!: React.ContextType<typeof TodoContext>;
  render() {
    return (
      <div className='modal-overlay modal-overlay-active'>
        <div className='todomodal padding-default mx-auto'>
          <div className='todomodal_inputs'>
            <div className=' my-1'>
              <input
                type='text'
                name='title'
                className='todomodal_input'
                placeholder='Add a title'
                value={this.state.title}
                onChange={(e: any) => {
                  this.setState({ title: e.target.value });
                }}
              />
            </div>
            {this.state.errors.titleError && (
              <p className='form-error-text'>{this.state.errors.titleError}</p>
            )}
            <div className='my-1'>
              <textarea
                className='todomodal_input'
                placeholder='Add a description'
                value={this.state.description}
                onChange={(e) => this.setState({ description: e.target.value })}
              />
            </div>
            {this.state.errors.descriptionError && (
              <p className='form-error-text'>
                {this.state.errors.descriptionError}
              </p>
            )}
            <div>
              <div className='d-flex justify-content-between my-2'>
                <div
                  className='btn toastmodal_btn'
                  onClick={() => this.props.onClose()}
                >
                  <p>cancel</p>
                </div>
                <div className='btn toastmodal_btn' onClick={this.handleSubmit}>
                  <p className='toastmodal_btn_text' data-testid='todoModal'>
                    {this.props.actionBtnTitle}
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

TodoModal.contextType = TodoContext;
