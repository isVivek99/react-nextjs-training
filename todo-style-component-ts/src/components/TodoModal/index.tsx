import React, { useState, useEffect } from 'react';
import { useTodoContext, actions } from '../../store';
import './styles.scss';
import styled from 'styled-components';

interface TodoModalProps {
  onClose: () => unknown;
  setTodos?: any;
  actionBtnTitle: string;
  todo?: any;
  newTodo: boolean;
  index?: number;
}

const StyledTodoModal = styled.div`
  width: 40%;
  border: 1px solid #ccc;
  border-radius: 1rem;
  background-color: #fff;
`;

const StyledInput = styled.input`
  border: 1px solid #ccc;
  border-radius: 0.5rem;
  font-size: 1.2rem;
  padding: 0.5rem;
  min-width: 90%;
  max-width: 90%;
  max-height: 400px;
  padding: 1rem;
`;

const StyledInputTextArea = styled.textarea`
  border: 1px solid #ccc;
  border-radius: 0.5rem;
  font-size: 1.2rem;
  padding: 0.5rem;
  min-width: 90%;
  max-width: 90%;
  max-height: 400px;
  padding: 1rem;
`;

const StyledTodoModalInputs = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  padding: 1rem;
  padding-bottom: 0;
  min-height: calc(100% - 20px);
  min-width: calc(100% - 20px);
  border-radius: 1rem;
  background-color: #fff;
`;

const ModalBtn = styled.div`
  background-color: #282c34;
  color: #fff;
  padding: 0.5rem;
  border-radius: 4px;
`;

const TodoModal = ({
  onClose,
  setTodos,
  actionBtnTitle,
  todo,
  newTodo = true,
  index,
}: TodoModalProps) => {
  const { todos, dispatchTodo } = useTodoContext();
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');

  const [errors, setErrors] = useState({
    title: '',
    description: '',
  });

  const handleSubmit = (e: any) => {
    e.preventDefault();

    setErrors({
      title: '',
      description: '',
    });

    //* form validation
    if (title.trim() === '' && description.trim() === '') {
      setErrors((prev) => ({
        ...prev,
        title: 'Title is required',
        description: 'Description is required',
      }));
    } else if (title.trim() === '') {
      setErrors((prev) => ({ ...prev, title: 'Title is required' }));
    } else if (description.trim() === '') {
      setErrors((prev) => ({
        ...prev,
        description: 'Description is required',
      }));
    } else {
      if (newTodo) {
        dispatchTodo({
          type: actions.addTodo,
          payload: { title, description },
          id: Math.floor(Math.random() * 100),
        });
        onClose();
      } else {
        setTodos((prev: any) => {
          let tempArr = [...prev];
          tempArr.splice(index ? index : 0, 1, { title, description });
          return tempArr;
        });
        onClose();
      }
    }
  };

  useEffect(() => {
    setTitle(todo ? todo.title : '');
    setDescription(todo ? todo.description : '');
  }, [todo]);

  return (
    <div className='modal-overlay modal-overlay-active'>
      {/* <div className='todomodal padding-default mx-auto'> */}
      <StyledTodoModal>
        <StyledTodoModalInputs>
          <div className=' my-1'>
            <StyledInput
              type='text'
              className='todomodal_input'
              placeholder='Add a title'
              value={title}
              onChange={(e) => setTitle(e.target.value)}
            />
          </div>
          {errors.title && <p className='form-error-text'>{errors.title}</p>}
          <div className='my-1'>
            <StyledInputTextArea
              className='todomodal_input'
              placeholder='Add a description'
              value={description}
              onChange={(e) => setDescription(e.target.value)}
            />
          </div>
          {errors.description && (
            <p className='form-error-text'>{errors.description}</p>
          )}
          <div>
            <div className='d-flex justify-content-between my-2'>
              <ModalBtn className='btn' onClick={() => onClose()}>
                <p>cancel</p>
              </ModalBtn>
              <ModalBtn className='btn' onClick={handleSubmit}>
                <p className='toastmodal_btn_text' data-testid='todoModal'>
                  {actionBtnTitle}
                </p>
              </ModalBtn>
            </div>
          </div>
        </StyledTodoModalInputs>
      </StyledTodoModal>
      {/* </div> */}
    </div>
  );
};

export default TodoModal;
