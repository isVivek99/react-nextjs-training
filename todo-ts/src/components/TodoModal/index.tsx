import React, { useState, useEffect } from 'react';
import './styles.scss';

interface TodoModalProps {
  onClose: () => unknown;
  setTodos?: any;
  actionBtnTitle: string;
  todo?: any;
  newTodo: boolean;
  index?: number;
}

const TodoModal = ({
  onClose,
  setTodos,
  actionBtnTitle,
  todo,
  newTodo = true,
  index,
}: TodoModalProps) => {
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');

  const [errors, setErrors] = useState({
    title: '',
    description: '',
  });

  const handleSubmit = (e: any) => {
    e.preventDefault();

    setErrors({
      title: '',
      description: '',
    });

    //* form validation
    if (title.trim() === '' && description.trim() === '') {
      setErrors((prev) => ({
        ...prev,
        title: 'Title is required',
        description: 'Description is required',
      }));
    } else if (title.trim() === '') {
      setErrors((prev) => ({ ...prev, title: 'Title is required' }));
    } else if (description.trim() === '') {
      setErrors((prev) => ({
        ...prev,
        description: 'Description is required',
      }));
    } else {
      if (newTodo) {
        setTodos((prev: any) => [...prev, { title, description }]);
        onClose();
      } else {
        setTodos((prev: any) => {
          let tempArr = [...prev];
          tempArr.splice(index ? index : 0, 1, { title, description });
          return tempArr;
        });
        onClose();
      }
    }
  };

  useEffect(() => {
    setTitle(todo ? todo.title : '');
    setDescription(todo ? todo.description : '');
  }, [todo]);

  return (
    <div className='modal-overlay modal-overlay-active'>
      <div className='todomodal padding-default mx-auto'>
        <div className='todomodal_inputs'>
          <div className=' my-1'>
            <input
              type='text'
              className='todomodal_input'
              placeholder='Add a title'
              value={title}
              onChange={(e) => setTitle(e.target.value)}
            />
          </div>
          {errors.title && <p className='form-error-text'>{errors.title}</p>}
          <div className='my-1'>
            <textarea
              className='todomodal_input'
              placeholder='Add a description'
              value={description}
              onChange={(e) => setDescription(e.target.value)}
            />
          </div>
          {errors.description && (
            <p className='form-error-text'>{errors.description}</p>
          )}
          <div>
            <div className='d-flex justify-content-between my-2'>
              <div className='btn toastmodal_btn' onClick={() => onClose()}>
                <p>cancel</p>
              </div>
              <div className='btn toastmodal_btn' onClick={handleSubmit}>
                <p className='toastmodal_btn_text'>{actionBtnTitle}</p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default TodoModal;
