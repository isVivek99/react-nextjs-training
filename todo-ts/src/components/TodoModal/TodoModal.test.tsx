import renderer from 'react-test-renderer';
import TodoModal from '.';

it('pops when clicked', () => {
  const component = renderer.create(
    <TodoModal
      onClose={() => console.log('close')}
      actionBtnTitle={'save'}
      newTodo={true}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
