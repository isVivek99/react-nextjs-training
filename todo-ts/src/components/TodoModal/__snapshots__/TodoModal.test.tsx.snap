// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pops when clicked 1`] = `
<div
  className="modal-overlay modal-overlay-active"
>
  <div
    className="todomodal padding-default mx-auto"
  >
    <div
      className="todomodal_inputs"
    >
      <div
        className=" my-1"
      >
        <input
          className="todomodal_input"
          onChange={[Function]}
          placeholder="Add a title"
          type="text"
          value=""
        />
      </div>
      <div
        className="my-1"
      >
        <textarea
          className="todomodal_input"
          onChange={[Function]}
          placeholder="Add a description"
          value=""
        />
      </div>
      <div>
        <div
          className="d-flex justify-content-between my-2"
        >
          <div
            className="btn toastmodal_btn"
            onClick={[Function]}
          >
            <p>
              cancel
            </p>
          </div>
          <div
            className="btn toastmodal_btn"
            onClick={[Function]}
          >
            <p
              className="toastmodal_btn_text"
              data-testid="todoModal"
            >
              save
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
